merkle/smt/hstar3.go

```diff
@@ -0,0 +1,183 @@
+// to the root(top) node(s) at the specified depth. The updates must have been
+// processed by Prepare.
+//
+// Note: The returned map could have bool value type, but []byte allows the
+// caller to reuse this map for filling in the hashes for the Update method.
+//
+// TODO(pavelkalinnikov): Return only tile IDs.
+func (h HStar3) GetReadSet(updates []NodeUpdate, top uint) map[tree.NodeID2][]byte {
+	ids := make(map[tree.NodeID2][]byte)
+	// For each node, add all its ancestors' siblings, down to the given depth.
+	for _, upd := range updates {
+		for id, d := upd.ID, h.depth; d > top; d-- {
+			pref := id.Prefix(d)
+			if _, ok := ids[pref]; ok {
+				// Delete the prefix node because its original hash does not contribute
+				// to the updates, so should not be read.
+				delete(ids, pref)
+				// All the upper siblings have been added already, so skip them.
+				break
+			}
+			sib := pref.Sibling()
+			ids[sib] = nil
+		}
+	}
+	return ids
+}
+
+// Update applies the given updates to a sparse Merkle tree. Requires a
+// previous Prepare invocation on the same updates slice. Returns an error if
+// any of the NodeStorage.Get calls does so, for example if a node is missing.
+//
+// Returns the slice of updates at the top level of the sparse Merkle tree
+// induced by the provided lower level updates. Typically it will contain only
+// one item for the root hash of a tile or a (sub)tree, but the caller may
+// arrange multiple subtrees in one call, in which case the corresponding
+// returned top-level updates will be sorted lexicographically by node ID.
+//
+// Note that Update invocations can be chained. For example, a bunch of HStar3
+// instances at depth 256 can return updates for depth 8 (in parallel), which
+// are then merged together and passed into another HStar3 at depth 8 which
+// computes the overall tree root update.
+//
+// For that reason, Update doesn't invoke NodeStorage.Set for the topmost level
+// nodes. If it did then chained Updates would Set the borderline nodes twice.
+//
+// Warning: This call modifies the updates slice in-place, so the caller must
+// ensure to not reuse it.
+func (h HStar3) Update(updates []NodeUpdate, top uint, ns NodeStorage) ([]NodeUpdate, error) {
+	for d := h.depth; d > top; d-- {
+		var err error
+		if updates, err = h.updateAt(updates, d, ns); err != nil {
+			return nil, fmt.Errorf("depth %d: %v", d, err)
+		}
+	}
+	return updates, nil
+}
+
+// updateAt applies the given node updates at the specified tree level.
+// Returns the updates that propagated to the level above.
+func (h HStar3) updateAt(updates []NodeUpdate, depth uint, ns NodeStorage) ([]NodeUpdate, error) {
+	// Apply the updates.
+	for _, upd := range updates {
+		ns.Set(upd.ID, upd.Hash)
+	}
+	// Calculate the updates that propagate to one level above.
+	newLen := 0
+	for i, ln := 0, len(updates); i < ln; i, newLen = i+1, newLen+1 {
+		sib := updates[i].ID.Sibling()
+		left := updates[i].Hash
```

Comment:
This could actually be the rightmost node at first, and only becomes left after the if/else below. It's fine when you notice the switcheroo with isLeftChild()... below, but until then it hurt my brain. Suggest leaving left and right both uninitialized here, and initialize both of them in the block below.

Reply:
Great idea. Done.